library/core/src/result.rs

```diff
@@ -853,6 +853,35 @@ impl<T, E> Result<T, E> {
             Err(e) => Err(op(e)),
         }
     }
+    
+    /// Returns `Err(e)` if the result is `Err(e)` (where `e` is the wrapped error),
+    /// otherwise calls `predicate` with the wrapped value and returns:
+    ///
+    /// - `Ok(t)` if `predicate` returns `Ok(())` (where `t` is the wrapped value), and
+    /// - `Err(e)` if `predicate` returns `Err(e)`.
```

Comment:
Why would one want to do `option.filter` rather than `option.and_then(|t| predicate(&t).then(|| t))`?

Reply:
False analogy. 

`{Option, Iterator}::filter` only needs a boolean result, so `option.filter(|t| t != 0)` works well. 

But `Result::check` requires user to supply the error value. This makes `Result::check` not "generic" since every "checker" has to adapt to the result error itself. You can't write `result.check(|t| t != 0)`. You must write 

```rust
result.check(|t| (t != 0).then_some(()).ok_or(Zero))
```

which is not too far away from 

```rust
result.and_then(|t| (t != 0).then_some(t).ok_or(Zero))
```

Reply:
We still maintain that we designed/intended it to be analogous to `Option::filter`. Specifically, it's meant to be the most straightforward adaptation of the `Option::filter` semantics onto `Result`. The only difference is that `Result` requires providing an error/reason when you discard the value.

The analogy can't be false - we literally designed it that way.

Reply:
you're ignoring my points that using the current `Result::check` is as complicated as `Result::and_then`, unlike `Option::filter` vs `Option::and_then`.

it is also not the single way to provide the error. you could do it like this:


```rust
result.check_or(|t| t != 0, Zero)
result.check_or_else(|t| t != 0, || Zero)
```

(and no i'm not saying this two-arg form is a good idea either)